components/executor/executor_impl.go

```diff
@@ -86,48 +96,81 @@ func (e *executorImpl) Delete(jobId string) error {
 	return ErrJobNotFound
 }
 
-// JobCh returns the channel used to receive jobs.
-func (e *executorImpl) JobCh() chan *jobmodels.Job {
-	return e.jobCh
-}
-
 // Close closes the executor and waits for all the jobs to finish executing.
 func (e *executorImpl) Close() {
 	e.rw.Lock()
 	defer e.rw.Unlock()
 
 	e.isClosed = true
 
-	// Wait for all the jobs to execute
-	for len(e.jobs) > 0 {
-		time.Sleep(1 * time.Second)
-	}
+	// Stop dispatcher and wait for it to stop
+	e.stopDispatcher()
+	e.wgDispacther.Wait()
 
+	// close dispatcher channel
+	// so that executor go-routines can terminate
 	close(e.jobCh)
 }
 
-func (e *executorImpl) scheduleJob(jobId string, version int, triggerTime time.Time) {
-	time.AfterFunc(time.Until(triggerTime), func() {
+func (e *executorImpl) nextTick() {
+	var now = time.Now()
+	for {
+		jentry := e.jobQueue.nextJob()
+		if jentry == nil {
+			return // We ran out of jobs
+		}
+		triggerTime := getTriggerTime(jentry.job)
+		if now.After(triggerTime) {
+			// Adding job back to the scheduler queue as it's ahead of current time
+			e.jobQueue.addJob(jentry)
+			return // We have processed jobs till current tick
+		}
 		e.rw.Lock()
-		defer e.rw.Unlock()
+		jobId := jentry.job.ID
 		entry, ok := e.jobs[jobId]
 		if !ok {
-			return // might be executed already
+			e.rw.Unlock()
+			continue // might be executed already
 		}
 
 		if entry.deleted {
 			// skip execution, job needs to be deleted
 			// delete from map
 			delete(e.jobs, jobId)
-		} else if entry.version == version {
+			e.rw.Unlock()
+			continue
+		} else if entry.version == jentry.version {
 			// latest job version
 			delete(e.jobs, jobId)
-
+			e.rw.Unlock()
 			e.jobCh <- entry.job
+		} else {
+			e.rw.Unlock()
+			continue
+		}
+	}
+}
+
+// startDispatcher is starts consuming jobs as per their trigger time
+// ctx is used to terminate tickering
+// it doesn't stop previous tickers jobs being dispatched if there are any
+func (e *executorImpl) startDispatcher(ctx context.Context) {
+	e.wgDispacther.Add(1)
+	go func() {
+		defer e.wgDispacther.Done()
+		var ticker = time.NewTicker(time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				e.nextTick()
```

Comment:
Since we are using time.After func, It will be calculating the delta for milliseconds. Should we sleep for few ms ?

Reply:
Also, on `ctx.Done()` we are returning from the routine. What about the pending jobs. Should we check for them before exiting?

Reply:
We are no longer using time.AfterFunc